### PR TITLE
Pagination branch

### PR DESCRIPTION
**Pull request by Divo123-tech**
## Things I did
**Page and Limit Validation**
I put the pagination function in the utils/utils.ts file with input validation that defaults a non-number value for skip to be 0, and limit to be 10

**Sorting**
Added sorting in the items.service.ts file and sorted by the name's alphabetical order

## Things to Review
1.the getPagination function in the utils/utils.ts file
2.the sorting in the items.service.ts file

### CODE REVIEW

File: src/controller/items.controller.ts [hunk 1]
```diff
@@ -1,5 +1,41 @@
-import TemplateController from './template.controller';
-import itemModel, { ItemInterface } from '../models/items.model'; // Import itemModel and IItem interface
+import itemService from "../service/item.service";
+import { Request, Response } from 'express';
+
+class ItemController {
+  static async getAll(req: Request, res: Response): Promise<void> {
+    try {
+        const {page, limit} = req.query
+        const skip: number = (Number(page) - 1) * Number(limit)
```
Comment (yharsono) [inline src/controller/items.controller.ts:8]: what if page and limit is not part of the query?
or what if there is only limit, but no page? or vice versa?

File: src/service/item.service.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        }
+    }
+    return false; // Return false if kode doesn't exist or if k is not provided
+}
+
+
+const getAll = async(skp: number,lmt : number) : Promise<Array<ItemInterface>> => {
+    
+    try {
+        return await itemModel.find({isActive: true}).skip(skp || 0).limit(lmt || 10);
```
Comment (yharsono) [inline src/service/item.service.ts:17]: you should sort this by something to make sure you always get consistent result. 
Its better to also tell the FE the total counts of the actual list, so they know how many pages they can show

**Review by yharsono — CHANGES_REQUESTED** [review]

**Review by yharsono — CHANGES_REQUESTED** [review]

File: src/service/item.service.ts [hunk 3]
```diff
@@ -1,65 +1,67 @@
...
+  limit: number
+): Promise<Array<ItemInterface>> => {
+  try {
+    const items = await itemModel.aggregate<ItemInterface>([
+      { $match: { isActive: true } },
+      { $sort: { name: 1 } },
+      { $skip: skip },
+      { $limit: limit },
+    ]);
+    return items;
```
Comment (yharsono) [inline src/service/item.service.ts:23]: you are still not returning the count of all the items

File: src/service/item.service.ts [hunk 4]
```diff
@@ -1,65 +1,67 @@
...
-        // return newItem;
-        return await itemModel.create(newItem);
-    }
-    catch (err: any) {
-        throw new Error(err)
-    }
-}
+const add = async (newItem: ItemInterface): Promise<ItemInterface> => {
+  try {
+    // return newItem;
```
Comment (yharsono) [inline src/service/item.service.ts:31]: nit: remove unused code

File: src/utils/utils.ts [hunk 5]
```diff
@@ -1,21 +1,16 @@
...
-    };
-  };
+  page: number;
+  limit: number;
+}
+export const getPagination = (query: QueryInterface) => {
+  //if limit is not given as a specific number use 10 as the default value
+  const limit: number = query.limit || 10;
+  const page: number = query.page || 1;
+  //query validation implemented in this function
```
Comment (yharsono) [inline src/utils/utils.ts:9]: what query validation?

File: src/utils/utils.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+export interface QueryInterface {
```
Comment (yharsono) [inline src/utils/utils.ts:1]: this is PaginationInterface, not QueryInterface

File: src/utils/utils.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+  page: number;
+  limit: number;
+}
+export const getPagination = (query: QueryInterface) => {
+  //if limit is not given as a specific number use 10 as the default value
+  const limit: number = query.limit || 10;
+  const page: number = query.page || 1;
+  //query validation implemented in this function
+  let skip: number = (page - 1) * limit; // Skip can be a number
```
Comment (yharsono) [inline src/utils/utils.ts:10]: const skip.

**Review by yharsono — CHANGES_REQUESTED** [review]

File: src/service/item.service.ts (hunk 3 above)
Comment (Divo123-tech) [reply src/service/item.service.ts:23]: should I make a new interface to return this kind of object?
{total:  number, data: items[]}

File: src/service/item.service.ts [hunk 8]
```diff
@@ -1,59 +1,66 @@
...
+  limit: number
+): Promise<{ totalCount: number; data: ItemInterface[] }> => {
+  try {
+    const items = await itemModel.aggregate<ItemInterface>([
+      { $match: { isActive: true } },
+      { $sort: { name: 1 } },
+      { $skip: skip },
+      { $limit: limit },
+    ]);
+    return { totalCount: items.length, data: items };
```
Comment (yharsono) [inline src/service/item.service.ts:23]: you cant just do items.lenght, cause its already been "limited" jdi this count will always return what the limit is, unless the amounts of items are smaller than the limit.

Check my previous comment on using $facet

**Review by yharsono — CHANGES_REQUESTED** [review]

**Comment by Divo123-tech** [comment]
<h3>Added Multiple Aggregation to getAll function</h3>
used .facet() to avoid running multiple queries and return an object with the total number of items in the list that are active, and all the items according to the skip, limit and sort filters.

<h3>Accidentally Merged services-branch to main </h3>
Merged the changes of services-branch to this branch so the changes can be reviewed
<h4>Adding new item interface for the front end.</h4>
 Added this interface for the add and edit functions to use as inputs to avoid front end editing or adding certain properties.
<h4>Used .save() to activate validation</h4>
removed doesKodeExist as the function became unnecesarry after using .save()
<h4>Used Object methods like .assign and .definePropertry() </h4>
Object.assign() was to update the item, and .defineProperty() was to ensure that isActive can only be made false by delete not edit or add.
<h4>Changed delete route </h4>
Changed delete from accepting id in the body to using the request params.

File: src/service/item.service.ts [hunk 9]
```diff
@@ -33,42 +49,42 @@ const add = async (newItem: FEItemInterface): Promise<ItemInterface> => {
...
-          throw new Error('Item not found');
-        }
-        //adds isActive property to newItem and if FrontEnd makes isActive into false
-        //it overrides it to true
-        Object.defineProperty(item, "isActive", {
-          value: true,
-        });
+    //adds isActive property to newItem and if FrontEnd makes isActive into false
+    //it overrides it to true
+    Object.defineProperty(item, "isActive", {
```
Comment (yharsono) [inline src/service/item.service.ts:72]: I dont think we need to worry about this anymore now that we have the FEItemInterface. 
And you could just do: `itemToUpdate.isActive = true` to always set it to true

File: src/service/item.service.ts [hunk 10]
```diff
@@ -33,42 +49,42 @@ const add = async (newItem: FEItemInterface): Promise<ItemInterface> => {
...
-    }
-    catch (err: any) {
-        throw new Error(err)
+const edit = async (
+  item: FEItemInterface,
+  id: string
+): Promise<FEItemInterface> => {
+  try {
+    const itemToUpdate = await itemModel.findById(id);
+    if (!itemToUpdate) {
```
Comment (yharsono) [inline src/service/item.service.ts:67]: you also need to check if `!itemToUpdate.isActive`, then you want to throw the item not found error, cause the item has been deleted

File: src/service/item.service.ts [hunk 11]
```diff
@@ -33,42 +49,42 @@ const add = async (newItem: FEItemInterface): Promise<ItemInterface> => {
...
-        });
+    //adds isActive property to newItem and if FrontEnd makes isActive into false
+    //it overrides it to true
+    Object.defineProperty(item, "isActive", {
+      value: true,
+    });
 
-        //assigns the ite passed to the item to be updated
-        Object.assign(itemToUpdate, item)
+    //assigns the ite passed to the item to be updated
```
Comment (yharsono) [inline src/service/item.service.ts:76]: nit: typo in the comment

**Review by yharsono — CHANGES_REQUESTED** [review]

**Review by yharsono — APPROVED** [review]

**Pull request merged**